pkg/api/convertertoagentpoolonlyapi.go

```diff
@@ -264,3 +266,63 @@ func convertVLabsAgentPoolOnlyCertificateProfile(vlabs *vlabs.CertificateProfile
 	api.KubeConfigCertificate = vlabs.KubeConfigCertificate
 	api.KubeConfigPrivateKey = vlabs.KubeConfigPrivateKey
 }
+
+func isAgentPoolOnlyCluster(version string, contents []byte) bool {
+	switch version {
+	case v20170831.APIVersion:
+		return true
+	case vlabs.APIVersion:
+		return isAgentPoolOnlyClusterFromProperties(contents, vlabsCountDistinctiveProperties)
+	default:
+		return false
+	}
+}
+
+func isAgentPoolOnlyClusterFromProperties(contents []byte, countDistinctiveProperties func(map[string]interface{}) distinctivePropertyCounts) bool {
+	properties, propertiesPresent := propertiesAsMap(contents)
+	if !propertiesPresent {
+		return false
+	}
+	counts := countDistinctiveProperties(properties)
+	return counts.AgentPoolOnlyClusterCount > counts.FullClusterCount
+}
+
+func propertiesAsMap(contents []byte) (map[string]interface{}, bool) {
+	var raw interface{}
+	json.Unmarshal(contents, &raw)
+	jsonMap := raw.(map[string]interface{})
+	properties, propertiesPresent := jsonMap["properties"]
+	if !propertiesPresent {
+		return nil, false
+	}
+	return properties.(map[string]interface{}), true
+}
+
+type distinctivePropertyCounts struct {
+	FullClusterCount          int
+	AgentPoolOnlyClusterCount int
+}
+
+func vlabsCountDistinctiveProperties(properties map[string]interface{}) distinctivePropertyCounts {
+	// The "vlabs" apiVersion is used for both full clusters and agent-only.
+	// To tell which this is, look for properties that differ between a full
+	// cluster .Properties and an AP-only cluster .Properties.  These are:
+	// - full should have orchestratorProfile, masterProfile
+	//        may have extensionProfiles, aadProfile
+	// - AP   should have dnsPrefix
+	//        may have fqdn, kubernetesVersion, kubernetesRelease, accessProfiles
+	fullClusterCount := 0
+	agentPoolOnlyClusterCount := 0
+	for k := range properties {
+		switch k {
+		case "orchestratorProfile", "masterProfile", "extensionProfiles", "aadProfile":
+			fullClusterCount++
+		case "dnsPrefix", "fqdn", "kubernetesVersion", "kubernetesRelease", "accessProfiles":
+			agentPoolOnlyClusterCount++
+		}
+	}
+	return distinctivePropertyCounts{
```

Comment:
This can be removed as well.

Safer/guaranteed way is:
If MasterProfile != nil or it contains property MasterProfile  then it's not a agent pool only cluster. Relying on count could break the logic in future because e.g. extensionProfiles could be added to agentpoolonly cluster and accessProfiles could be added to full cluster.

Reply:
Thanks @amanohar for the advice and feedback. If we can use masterProfile alone then that's much nicer - it greatly simplifies the logic.  I've made that change; please let me know if you would like to see further changes.